c2rust-analyze/src/panic_detail.rs

```diff
@@ -0,0 +1,170 @@
+        warn!("discarding old panic detail: {:?}", old);
+    }
+}
+
+/// Get the [`PanicDetail`] of the most recent panic.  This clears the internal storage, so if this
+/// is called twice in a row without an intervening panic, the second call always returns `None`.
+pub fn take_current() -> Option<PanicDetail> {
+    CURRENT_PANIC_DETAIL.with(|cell| cell.take())
+}
+
+/// Get the [`PanicDetail`] of the most recent panic; if it's not available, build a placeholder
+/// `PanicDetail` from the panic payload `e`.  This is useful in conjunction with
+/// [`std::panic::catch_unwind`].
+pub fn catch(e: &(dyn Any + Send + 'static)) -> PanicDetail {
+    take_current().unwrap_or_else(|| {
+        let msg = panic_to_string(e);
+        warn!("missing panic detail; caught message {:?}", msg);
+        PanicDetail::new(msg)
+    })
+}
+
+/// Crude heuristic to guess the first interesting location in a [`Backtrace`], skipping over
+/// helper functions, wrappers, and panic machinery.  The resulting location is used in the summary
+/// message produced by [`PanicDetail::to_string_short`].
+fn guess_relevant_loc(bt: &Backtrace) -> Option<String> {
+    for frame in bt.frames() {
+        for symbol in frame.symbols() {
+            let name = match symbol.name() {
+                Some(x) => x.to_string(),
+                None => continue,
+            };
+            if name.starts_with("c2rust_analyze::dataflow")
+                || name.starts_with("c2rust_analyze::borrowck")
+                || name.starts_with("c2rust_analyze::rewrite")
+                || name.contains("type_of_rvalue")
+                || name.contains("TypeOf")
+                || name.contains("lty_project")
+            {
+                let filename_str = match symbol.filename() {
+                    Some(x) => x.display().to_string(),
+                    None => "[unknown]".to_string(),
+                };
+                return Some(format!(
+                    "{} @ {}:{}:{}",
+                    name,
+                    filename_str,
+                    symbol.lineno().unwrap_or(0),
+                    symbol.colno().unwrap_or(0)
+                ));
+            }
+        }
+    }
+    None
+}
+
+fn panic_to_string(e: &(dyn Any + Send + 'static)) -> String {
+    match e.downcast_ref::<&'static str>() {
+        Some(s) => return s.to_string(),
+        None => {}
+    }
+
+    match e.downcast_ref::<String>() {
+        Some(s) => return (*s).clone(),
+        None => {}
+    }
+
+    format!("unknown error: {:?}", e.type_id())
+}
+
+thread_local! {
+    static CURRENT_SPAN: Cell<Span> = Cell::new(DUMMY_SP);
+}
+
+pub struct CurrentSpanGuard {
+    old: Span,
+}
+
+impl Drop for CurrentSpanGuard {
+    fn drop(&mut self) {
+        CURRENT_SPAN.with(|cell| cell.set(self.old));
+    }
+}
+
+/// Set the current span.  Returns a guard that will reset the current span to its previous value
+/// on drop.
+pub fn set_current_span(sp: Span) -> CurrentSpanGuard {
```

Comment:
This is very neat.

Would `enter_current_span` or just `enter_span` be a bit clearer (similar to Python's `__enter__` and `__exit__`)?

Reply:
Functions named `enter_foo` or `with_foo` usually take a closure and set up some kind of context for the duration of that closure.  Here we're using a guard because it's less obtrusive at call sites and we're not likely to screw up the nesting.